test/test_prototype_datasets_utils.py

```diff
@@ -48,6 +51,197 @@ def test_read_flo(tmpdir):
+
+    def test_priority_decompressed_gt_raw(self, tmp_path):
+        # We don't need to actually compress here. Adding the suffix is sufficient
+        self._make_file(tmp_path, content="raw_sentinel", name="file.txt.gz")
+        file = self._make_file(tmp_path, content="decompressed_sentinel", name="file.txt")
+
+        resource = self.DummyResource(file_name=file.name)
+
+        dp = resource.load(tmp_path)
+        path, buffer = next(iter(dp))
+
+        assert path == str(file)
+        assert buffer.read().decode() == "decompressed_sentinel"
+
+    def test_priority_extracted_gt_decopressed(self, tmp_path):
+        archive, _ = self._make_tar(tmp_path, remove=False)
+
+        resource = self.DummyResource(file_name=archive.name)
+
+        dp = resource.load(tmp_path)
+        # If the archive had been selected, this would be a `TarArchiveReader`
+        assert isinstance(dp, FileOpener)
+
+    def test_download(self, tmp_path):
+        file_name = "file.txt"
+        content = "sentinel"
+
+        resource = self.DummyResource(
+            file_name=file_name,
+            download_fn=lambda resource, root: self._make_file(root, content=content, name=resource.file_name),
+        )
+
+        dp = resource.load(tmp_path)
+        data = list(dp)
+        assert len(data) == 1
+
+        path, buffer = data[0]
+        assert path == str(tmp_path / file_name)
+        assert buffer.read().decode() == content
+
+    def test_preprocess_decompress(self, tmp_path):
+        file_name = "file.txt.gz"
+        content = "sentinel"
+
+        def download_fn(resource, root):
+            file = root / resource.file_name
+            with gzip.open(file, "wb") as fh:
+                fh.write(content.encode())
+            return file
+
+        resource = self.DummyResource(file_name=file_name, preprocess="decompress", download_fn=download_fn)
+
+        dp = resource.load(tmp_path)
+        data = list(dp)
+        assert len(data) == 1
+
+        path, buffer = data[0]
+        assert path == str(tmp_path / file_name).replace(".gz", "")
+        assert buffer.read().decode() == content
+
+    def test_preprocess_extract(self, tmp_path):
+        files = None
+
+        def download_fn(resource, root):
+            nonlocal files
+            archive, files = self._make_tar(root, name=resource.file_name)
+            return archive
+
+        resource = self.DummyResource(file_name="folder.tar", preprocess="extract", download_fn=download_fn)
+
+        dp = resource.load(tmp_path)
+        assert files is not None, "`download_fn()` was never called"
+        assert isinstance(dp, FileOpener)
+
+        actual = {path: buffer.read().decode() for path, buffer in dp}
+        expected = {
+            path.replace(resource.file_name, resource.file_name.split(".")[0]): content
+            for path, content in files.items()
+        }
+        assert actual == expected
```

Comment:
Of course there are other ways, but I'm not sure they are easier. We can't create the archive upfront, because in that case we would never trigger the preprocessing. One option would be to create the data upfront in a temporary directory and move it to `root` inside the download function similar to what we are doing with the actual resource loading in our dataset tests (#6010).

Given that you were ok with `nonlocal` in https://github.com/pytorch/vision/pull/5990#discussion_r873553403, I don't think it will be much simpler. You choose.